### PR TITLE
obb: init at 0.0.1

### PR DESCRIPTION
**Pull request by willcohen**
###### Motivation for this change

Add initial release of [obb](https://github.com/babashka/obb), a Mac scripting tool built off of [babashka](https://babashka.org/).

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/clojure/obb.nix
Comment (legendofmiracles) [inline pkgs/development/interpreters/clojure/obb.nix:21]: There is [`testVersion`](https://github.com/NixOS/nixpkgs/blob/de49d559382ea05a304589b6fc13c9be68351810/pkgs/build-support/trivial-builders.nix#L724-L755) which can be used here instead. See https://github.com/NixOS/nixpkgs/commit/255c2849248e331d7f0e56a041b80e0014a19177 for an example.


The simple 1 + 2 should also be a [package test](https://nixos.org/manual/nixpkgs/stable/#ssec-inline-package-tests-writing)

File: pkgs/development/interpreters/clojure/obb.nix
Comment (legendofmiracles) [inline pkgs/development/interpreters/clojure/obb.nix:12]: I don't think this is needed?

File: pkgs/development/interpreters/clojure/obb.nix
Comment (legendofmiracles) [inline pkgs/development/interpreters/clojure/obb.nix:10]: [fetchFromGitHub](https://nixos.org/manual/nixpkgs/stable/#fetchfromgithub) is preferred when we're only fetching the source code

File: pkgs/development/interpreters/clojure/obb.nix
Comment (legendofmiracles) [inline pkgs/development/interpreters/clojure/obb.nix:1]:
```suggestion
{ lib, stdenv, fetchurl }:
```
Unused?

File: maintainers/maintainer-list.nix
Comment (legendofmiracles) [inline maintainers/maintainer-list.nix:12679]: Different commit: `maintainers: add willcohen`

File: pkgs/development/interpreters/clojure/obb.nix
Comment (willcohen) [reply pkgs/development/interpreters/clojure/obb.nix:10]: The github release is actually a binary. I suppose I actually should be modifying the derivation to compile from source instead and improve reproducibility.

File: pkgs/development/interpreters/clojure/obb.nix
Comment (willcohen) [reply pkgs/development/interpreters/clojure/obb.nix:51]: Version test doesn't even work with 0.0.1, so removing that for now in addition to fixing the package test

**Comment by willcohen** [comment]
@legendofmiracles many thanks for the review, changes incorporated with 7d0a4ea!

**Comment by willcohen** [comment]
@thiagokokada I included you as maintainer since you're on babashka too, but don't feel obliged. I just felt a little strange adding myself as the only maintainer for a new formula!

File: pkgs/development/interpreters/clojure/obb.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/clojure/obb.nix:28]:
```suggestion
  configurePhase = ''
    runHook preConfigure

    mkdir -p .m2
    substituteInPlace deps.edn --replace ':paths' ':mvn/local-repo "./.m2" :paths'
    substituteInPlace bb.edn --replace ':paths' ':mvn/local-repo "./.m2" :paths'
    echo deps.edn

    runHook postConfigure
  '';
```

File: pkgs/development/interpreters/clojure/obb.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/clojure/obb.nix:38]:
```suggestion
    runHook preBuild

    export DEPS_CLJ_TOOLS_DIR=${clojure}
    export DEPS_CLJ_TOOLS_VERSION=${clojure.version}
    mkdir -p .gitlibs
    mkdir -p .cpcache
    export GITLIBS=.gitlibs
    export CLJ_CACHE=.cpcache

    bb build

    runHook postBuild
```

File: pkgs/development/interpreters/clojure/obb.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/clojure/obb.nix:46]:
```suggestion
    runHook preInstall

    install -Dm755 "out/bin/obb" "$out/bin/obb"

    runHook postInstall
```

File: pkgs/development/interpreters/clojure/obb.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/clojure/obb.nix:67]: I don't have a Mac so I am not interested in maintaining this package, sorry.
```suggestion
```

Comment (willcohen) [reply pkgs/development/interpreters/clojure/obb.nix]: Okay! Thanks for reviewing!

File: pkgs/development/interpreters/clojure/obb.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/clojure/obb.nix:70]: I think it is better to use `installCheckPhase` instead:

```suggestion
  doInstallCheck = true;
  installCheckPhase = ''
    [ $($out/bin/obb -e '(+ 1 2)' = '3' ]
  '';
```

This way, you don't need to have the derivation depending on itself.

**Comment by willcohen** [comment]
@thiagokokada review comments adopted in latest commit. thanks again!

@legendofmiracles updated test now actually tests for a working install -- similar to with [macvim's path issues](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/pkgs/applications/editors/vim/macvim.nix#L23), `osascript` needed to be symlinked into `$out` for the test to pass, since it wasn't being found otherwise (though it still works on darwin for the end user). also moved it back to `installCheckPhase` per @thiagokokada's second review to avoid the circular dependency, but now it actually tests correctly.

File: pkgs/development/interpreters/clojure/obb.nix
Comment (willcohen) [reply pkgs/development/interpreters/clojure/obb.nix]: Makes sense, thanks!

File: pkgs/development/interpreters/clojure/obb.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/clojure/obb.nix:11]: Just my TOC:
```suggestion
, runCommand }:
```

**Comment by thiagokokada** [comment]
I can at least give a try of this PR in my VM and see if it build/works.

File: pkgs/development/interpreters/clojure/obb.nix
Comment (willcohen) [reply pkgs/development/interpreters/clojure/obb.nix]: Good catch -- adjusted accordingly below

**Comment by willcohen** [comment]
Result of `nixpkgs-review pr 153496` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>obb</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 153496` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>obb</li>
  </ul>
</details>

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
Ran the example from README:

```
$ obb -e '(-> (js/Application "Safari") (.-documents) (aget 0) (.url))'
https://www.apple.com/
```

Seems to work fine.

**Comment by willcohen** [comment]
Thank you!

**Comment by legendofmiracles** [comment]
Some things I didn't have a chance to review since the last changes.
`src` should be fetchFromGitHub now that we aren't downloading a binary release. 

And I also disagree with moving the tests, the `passthru.tests` exists for a reason and many derivations rely on themselves (note that it's not a circular dependency). Although there is a case to be made that being forced to create a $out file makes things more ugly and unnecessarily complicated.
I didn't have a chance to see for my own why 0.0.1 wouldn't work with testVersion.

I do appreciate the work though, @willcohen. Will open another PR when I have the time.

**Comment by thiagokokada** [comment]
> Some things I didn't have a chance to review since the last changes. `src` should be fetchFromGitHub now that we aren't downloading a binary release.

Sure, this is something that I didn't consider. Can be fixed later though, and it is a minor issue.

> I disagree with moving the tests, the `passthru.tests` exists for a reason and many derivations rely on themselves (note that it's not a circular dependency). Although there is a case to be made that being forced to create a $out file makes things more ugly and unnecessarily complicated. I didn't have a chance to see for my own why 0.0.1 wouldn't work with testVersion.

What is the issue with `installCheckPhase` though? Legit question since I never saw any reviewer implicating with its usage.

Also, I know about `passthru.tests`, however AFAIK they're mostly used for when you have a test in `nixos/tests` and you want to also run those tests when the package is updated. Not really the case here though.

**Pull request merged**
